bittide/src/Clash/Hitl.hs

```diff
@@ -0,0 +1,289 @@
+          }
+      | (id_, dat) <- probes]
+    }
+
+-- | Convert a collection of named tests ("HitlTests") to a \"packed\" representation
+-- readable by our TCL test infrastructure. It will generate YAML/JSON that looks
+-- like:
+--
+-- > defaults:
+-- >   probes:
+-- >     probe_test_data: 0
+-- >
+-- > tests:
+-- >   testname1:
+-- >     targets:
+-- >       - id: 0
+-- >         probes:
+-- >           probe_test_data: <binary representation of 'a'>
+-- >       - id: 1
+-- >         probes:
+-- >           probe_test_data: <binary representation of 'a'>
+-- >       ...
+-- >   testname2:
+-- >     ...
+--
+packAndEncode :: forall a. BitPack a => HitlTests a -> LazyByteString.ByteString
+packAndEncode = encodePretty . toPacked
+
+-- | Whether a test has been completed, see 'hitlVio'.
+type Done = Bool
+
+-- | Whether a test has been completed successfully, see 'hitlVio'.
+type Success = Bool
+
+-- | Instantiate this VIO in a design you'd like to test with hardware in the
+-- loop. Its output is set to 'Nothing' if a test is not running, and will be
+-- set to 'Just' if it is.
+hitlVio ::
+  forall a dom .
+  ( KnownDomain dom
+  , BitPack a
+   ) =>
+  -- | Default value for @a@. This is an artifact of this VIO internally representing
+  -- the output value as two probes (\"valid\" and \"data\") to accomodate the
+  -- TCL infrastructure. Hence, the actual value of the default doesn't matter:
+  -- whenever it is output, this VIO will output 'Nothing'.
+  --
+  -- TODO: Allow use of 'errorX' in 'vioProbe'
+  a ->
+  P.Clock dom ->
+  -- | Should be asserted when a test is done. For sanity checking the TCL
+  -- infrastructure, this must be *deasserted* when a test is not running.
+  P.Signal dom Done ->
+  -- | When 'Done' is asserted, this signal indicates whether a test has been
+  -- completed successfully.
+  P.Signal dom Success ->
+  -- | Test values supplied by the VIO. Test modules should export a symbol
+  -- @tests :: HitlTests a@ that defines the data.
+  P.Signal dom (Maybe a)
+hitlVio dflt clk done success | 0 <- natToInteger @(BitSize a) =
+  -- XXX: This branch is a workaround for 'vioProbe' not handling zero-width
+  --      ports properly.
+  P.mux start (pure (Just dflt)) (pure Nothing)
+ where
+  start =
+    P.setName @"vioHitlt" $
+      vioProbe
+        ("probe_test_done" :> "probe_test_success" :> Nil)
+        ("probe_test_start" :> Nil)
+        False clk done success
+
+hitlVio dflt clk done success =
+  P.mux start (Just <$> dat) (pure Nothing)
+ where
+  (P.unbundle -> (start, dat)) =
+    P.setName @"vioHitlt" $
+      vioProbe
+        ("probe_test_done" :> "probe_test_success" :> Nil)
+        ("probe_test_start" :> "probe_test_data" :> Nil)
+        (False, dflt) clk done success
+
+-- | Instantiate this VIO in a design you'd like to test with hardware in the
+-- loop. Its output is set to 'True' if a test is not running, and will be
+-- set to 'False' if it is.
+hitlVio1 ::
```

Comment:
I dislike `name<n>` functions because you have to look up how they differentiate.
how about using `hitlVioWithConfig` and `hitlVio` or `hitlVioBool`? Any other choices are fine too imo.

Reply:
I like `hitlVioBool`